language/move-core/types/src/vm_status.rs

```diff
@@ -235,9 +235,26 @@ pub fn convert_prologue_runtime_error(status: VMStatus) -> VMStatus {
             };
             VMStatus::Error(new_major_status)
         }
-        status @ VMStatus::ExecutionFailure { .. }
-        | status @ VMStatus::Error(_)
+        status @ VMStatus::ExecutionFailure { .. } | status @ VMStatus::Executed => status,
```

Comment:
are we sending `ExecutionFailure` back? sorry I am getting confused here...

Reply:
ExecutionFailure and Executed are passed through. 
- Executed is obvious as to why (we want to be able to run the prologue in a success case)
- ExecutionFailure should be impossible given the definition, but I want to prevent any writeset nonsense from borking the VM here (i.e. the VM vomiting after the writeset caused the code to execution failure in some case)

Reply:
fine on `Executed` of course.
I don't understand `ExecutionFailure`. This is `prologue` handling, or is it? An error here should never charge the user, aren't we going to charge here if we get an `ExecutionFailure` during prologue?

Reply:
Yes. 

My worry here is that we fail on some arithmetic or something, and then by not charging, we are somewhat vulnerable to an attack. 

Additionally, I worry that some change in the prologue/epilogue (via a writeset) would *want* an execution error because of arithmetic or a missing resource or something similar. But then the VM would need to change to accommodate that writeset. 

@sblackshear @tzakian Any thoughts here? How likely is it that the prologue/epilogue could have a runtime error (div by zero, missing resource, etc)?

Reply:
The likelihood of the prologue running into an internal error like this should be none, and should be treated in a similar manner to an invariant violation in the VM (i.e., no charge). Especially for the prologue no charge should ever happen if it fails since at that point we don't "know" whether the sending account actually owns the account...

For the epilogue, I would likewise classify execution failures arising from here as an invariant violation. I would likewise push to not charge in such instances. But, I could be convinced otherwise here if anyone feels strongly a particular way. 

In general we've taken a pretty strict view of the prologue/epilogue of "it should _never_ abort with anything other than the precise set of abort conditions that are expected" anything else should be treated as an invariant violation since we are essentially straying in to undefined transaction behavior at that point.

Reply:
I agree with Tim. I think anything that doesn't fall into the expected behavior of the system should be a discard (in both prologue and epilogue).

Reply:
Thanks for the input!

- I'll mark execution failures for the prologue as invariant violations
- For the epilogue, I will split it into success/failure ones. The success one will need to be more flexible to be able to kick over to the failure one